documentation/proposals/Proposal - Load overrides.md

```diff
@@ -0,0 +1,157 @@
+# Summary
+Load overrides mean a way to override loading native addresses.
+This is useful in the following scenarios:
+- P/Invoke / statically linked scenarios, mobile.
+- Advanced Performance scenarios, such as pre-loading to improve AOT / linking / R2R
+
+While this API can be used by Hand, and is available for public use it is not designed to be particularly user friendly.
+It is designed with SilkTouch as the primary user. (See `PInvokeTableAttribute`)
+
+This proposal also somewhat defines the currently loose timeline of how addresses are loaded.
+
+
+# Contributors
+- Kai Jellinghaus, Maintainer (at time of writing), open source community.
+
+# Current Status
+- [x] Proposed
+- [ ] Discussed with API Review Board (ARB)
+- [ ] Approved
+- [ ] Implemented
+
+# Design Decisions
+There are two separate interfaces to allow maximum flexibility, but also to define exactly at which point slot information is lost.
+
+# Proposed API
+Note that this API may be changed in the future to use function pointers directly instead of IntPtrs.
+This deliberately doesn't use nint.
+```cs
+/// <summary>
+/// Defines a Load override, capable of loading some entrypoints by name.
+/// </summary>
+/// <remarks>
+/// Load overrides may not cache loads.
+/// Load overrides have to be thread safe.
+/// </remarks>
+public interface INameLoadOverride
+{
+    /// <summary>
+    /// The <see cref="INativeContext"/> used for loading
+    /// </summary>
+    INativeContext Context { set; }
+
+    /// <summary>
+    /// Attempt to load an entrypoint by name.
+    /// </summary>
+    /// <param name="entrypoint">The entrypoint name to load</param>
+    /// <param name="address">The address that was loaded</param>
+    /// <returns>
+    /// Whether the load was successful.
+    /// </returns>
+    bool TryLoad(string entrypoint, out IntPtr address);
+}
+
+/// <summary>
+/// Defines a Load override, capable of loading some entrypoints by slot.
+/// </summary>
+/// <remarks>
+/// Load overrides may not cache loads.
+/// Load overrides have to be thread safe.
+/// </remarks>
+public interface ISlotLoadOverride
+{
+    /// <summary>
+    /// The <see cref="INativeContext"/> used for loading
+    /// </summary>
+    INativeContext Context { set; }
+        
+    /// <summary>
+    /// Attempt to load an entrypoint by name.
+    /// </summary>
+    /// <param name="slot">The slot to load</param>
+    /// <param name="address">The address that was loaded</param>
+    /// <returns>
+    /// Whether the load was successful.
+    /// This will flush the cache / VTable.
+    /// </returns>
+    bool TryLoad(int slot, out IntPtr address);
+}
+```
+
+Not a critical change, but would make sense to make the core easier to understand:
+```diff
+- public interface INativeContext
++ public interface IAddressLoader
```

Comment:
Personally would like to vito on this name change, feel free to add docs to accomplish better understanding.

Reply:
As in, I'd rather not rename the class.

Reply:
I've added some more explanation, but I don't really mind this either, just figured it would fit in 🙂